lib/logstash/outputs/mongodb.rb

```diff
@@ -46,19 +46,22 @@ class LogStash::Outputs::Mongodb < LogStash::Outputs::Base
   # Mutex used to synchronize access to 'documents'
   @@mutex = Mutex.new
 
-  public
   def register
+    if @bulk_size > 1000
+      raise LogStash::ConfigurationError, "Bulk size must be lower than '1000', currently '#{@bulk_size}'"
+    end
+
     Mongo::Logger.logger = @logger
     conn = Mongo::Client.new(@uri)
     @db = conn.use(@database)
 
-    if @bulk_size > 1000
-      raise LogStash::ConfigurationError, "Bulk size must be lower than '1000', currently '#{@bulk_size}'"
-    end
+    @closed = Concurrent::AtomicBoolean.new(false)
     @documents = {}
-    Thread.new do
+
+    @bulk_thread = Thread.new(@bulk_interval) do |bulk_interval|
       loop do
-        sleep @bulk_interval
+        sleep(bulk_interval)
+
         @@mutex.synchronize do
```

Comment:
first time seeing this code, am surprised at an implementation-level mutex for database operations. I assume this is so we can share a document buffer across worker threads, but it effectively also constrains us to at most one mongo operation concurrently when we have multiple instances of the Mongo Output, even if those instances are in separate pipelines.

Reply:
Yes, did not want to change that because it is out-of-scope and risks changing behaviours in ways it is difficult for me to track. If a refactor of this is eventually needed we can revisit.